src/dbnode/storage/index/postings_list_cache.go

```diff
@@ -0,0 +1,253 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"sync"
+	"time"
+
+	"github.com/m3db/m3/src/m3ninx/postings"
+	"github.com/m3db/m3x/instrument"
+
+	"github.com/pborman/uuid"
+	"github.com/uber-go/tally"
+)
+
+// PatternType is an enum for the various pattern types. It allows us
+// separate them logically within the cache.
+type PatternType int
+
+const (
+	// PatternTypeRegexp indicates that the pattern is of type regexp.
+	PatternTypeRegexp PatternType = iota
+	// PatternTypeTerm indicates that the pattern is of type term.
+	PatternTypeTerm
+
+	reportLoopInterval = 10 * time.Second
+)
+
+// PostingsListCacheOptions is the options struct for the query cache.
+type PostingsListCacheOptions struct {
+	InstrumentOptions instrument.Options
+}
+
+// PostingsListCache implements an LRU for caching queries and their results.
+type PostingsListCache struct {
+	sync.Mutex
+
+	lru *postingsListLRU
+
+	size    int
+	opts    PostingsListCacheOptions
+	metrics *postingsListCacheMetrics
+
+	stopReporting chan struct{}
+}
+
+// NewPostingsListCache creates a new query cache.
+func NewPostingsListCache(size int, opts PostingsListCacheOptions) (*PostingsListCache, error) {
+	lru, err := newPostingsListLRU(size)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PostingsListCache{
+		lru:     lru,
+		size:    size,
+		opts:    opts,
+		metrics: newPostingsListCacheMetrics(opts.InstrumentOptions.MetricsScope()),
+	}, nil
+}
+
+// GetRegexp returns the cached results for the provided regexp query, if any.
+func (q *PostingsListCache) GetRegexp(
+	segmentUUID uuid.UUID,
+	pattern string,
+) (postings.List, bool) {
+	return q.get(
+		segmentUUID,
+		pattern,
+		PatternTypeRegexp)
+}
+
+// GetTerm returns the cached results for the provided term query, if any.
+func (q *PostingsListCache) GetTerm(
+	segmentUUID uuid.UUID,
+	pattern string,
+) (postings.List, bool) {
+	return q.get(
+		segmentUUID,
+		pattern,
+		PatternTypeTerm)
+}
+
+func (q *PostingsListCache) get(
+	segmentUUID uuid.UUID,
+	pattern string,
+	patternType PatternType,
+) (postings.List, bool) {
+	// No RLock because a Get() operation mutates the LRU.
```

Comment:
what's the contention on this thing like? could do a bunch of things to make it less expensive (e.g. approximate the lru, sample updates, etc)

Fine to do this as a follow up too.

Reply:
doesn't show up in flame graphs at all, and at the load we're currently handling I don't think its highly contended (100s of queries/s at peak, maybe looking back 100 blocks each is 10,000 locks per second which isn't much). Would rather keep implementation simple (already kind of sucks I had to specialize it) until it becomes an issue which I don't think will happen until we have 1-2 orders of magnitude improvement in general query performance

Reply:
wouldn't show up in the cpu flamegraphs necessarily - did you try the blocking?.

mainly curious if this is the cause of latency spikes you mentioned but we can investigate this separately. don't want to block the PR.

Reply:
ah...no I did not check blocking. I will get a blocking profile next time I do benchmarking and see if this is an issue. My guess is that its not the root cause of the latency spikes but I'd love to be wrong as this would be an easy fix lol

Reply:
Added an additional bullet point to #1365 to track